tests/testthat/test-utility-extra.R

```diff
@@ -0,0 +1,400 @@
+# Test utility helper functions against R counterparts
+
+# Pedestrian names
+nms = c('a1', 'a2', 'c1')
+
+# Current pedestrian
+p1 = matrix(c(0, 0), 1, 2)
+rownames(p1) = nms[1]
+
+# All pedestrians
+p2 = rbind(c(0, 0), c(0.5, 0.5), c(1, 1))
+rownames(p2) = nms
+
+# Pedestrian angle in front
+a_front = 90
+names(a_front) = nms[1]
+
+# Pedestrian angles not in front
+a_not_front = 135
+names(a_not_front) = nms[1]
+
+# Pedestrian angles vector
+a = c(90, 90, 90)
+names(a) = nms
+
+# Pedestrian velocities
+v = c(0.5, 0.5, 0.5)
+names(v) = nms
+
+# Goals
+P1 = matrix(c(2, 2, -1, -1), 2, 2)
+rownames(P1) = c('g1', 'g2')
+attr(P1, 'i') = 1
+
+# Pedestrian radius
+r = c(0.5, 0.5, 0.5)
+names(r) = nms
+
+# Pedestrian group
+group = c(2, 1, 2)
+names(group) = nms
+
+# Cell centres
+set.seed(23123)
+centres = matrix(rnorm(66), 33, 2)
+
+# Predicted pedestrian positions
+p_pred = rbind(c(0, 0), c(0.75, 0.75), c(1.25, 1.25))
+rownames(p_pred) = nms
+
+# Objects not occluding view between p1 and p2
+objects = list(
+  list(x = c(1, 0), y = c(0, 1)),
+  list(x = c(0, 1), y = c(1, 0))
+)
+
+# Objects occluding view between p1 and p2
+objects_occlude = list(
+  list(x = c(0.15, 0.35), y = c(0.15, 0.35))
+)
+
+# State list
+state = list(
+  p = p2,
+  a = c(45, 45, 45),
+  v = v,
+  r = r,
+  P = list(
+    P1, P1, P1
+  ),
+  group = group
+)
+names(state$a) = nms
+
+# Pos of current pedestrian in p2
+n = 1
+
+testthat::test_that("Destination angle computation works", {
+  ref = m4ma::destinationAngle_r(a_front, p1, P1)
+  res = m4ma::destinationAngle_rcpp(a_front, p1, P1)
+  testthat::expect_equal(res, ref)
+})
+
+testthat::test_that(
+  "Predicted distance works (when all pedestrians are seen in front)", {
+    ref = m4ma::predClose_r(n, p1, a_front,
+                            p2, r, centres, p_pred, objects)
+    res = m4ma::predClose_rcpp(n-1, p1, a_front,
+                               p2, r, centres, p_pred, objects)
+    testthat::expect_equal(res, ref)
+  }
+)
+
+testthat::test_that(
+  "Predicted distance works (when one or less pedestrians are occluded)", {
+    ref = m4ma::predClose_r(n, p1, a_front,
+                            p2, r, centres, p_pred, objects_occlude)
+    res = m4ma::predClose_rcpp(n-1, p1, a_front,
+                               p2, r, centres, p_pred, objects_occlude)
+    testthat::expect_equal(res, ref)
+    testthat::expect_equal(res, NULL)

```

Comment:
I am confused about line 101. I get that in some circumstances the output of predClose_rcpp must be NULL, but why these circumstances are different than the ones contemplated by predClose_r? 

What I mean, is that if both functions are indeed doing the same thing, than testthat::expect_equal(res, ref) should work also for circumstances when the output is NULL (_if_ both functions are the same). I find this a possible source of bug.

Would you mind elaborate here this decision?

Reply:
In this test, both the R and Rcpp implementations should always return NULL. I added this extra `expect_equal(res, NULL)` statement to highlight and enforce this requirement. That is, if this test fails, we know that both implementations are broken. Do you agree with this rationale?

Reply:
so, basically you are mixing two different tests. One that tests whether the result of function 1 is the same as function 2 and the other whether both functions work as expected (i.e., they are equal to NULL). I personally find the second test redundant, and I am not sure why you don't trust the r function to begin with. Again, do as you like, if you wish to keep this second test, in the context of developing a test suite that could be interpreted to anyone  (not only you as developer), maybe you may want to consider splitting this rationale in two separate test whereby you make explicit the goal of each single test?

Reply:
I see your point and separated the tests for `NULL` and `numeric(0)`. Indeed, otherwise it's not obvious if just the C++ function gives a different result than the R one or if both give a result other than `NULL`.